### PR TITLE
Support RBAC for cloud-controller-manager for local-up-cluster

### LINKED ISSUE #59943: RBAC for cloud-node-controller and pvl-controller not set up properly
when i run local-up-cluster with the cloud-controller-manager (pointing to openstack in my case), i see the following in my `cloud-controller-manager.log`
 
```
dims@bigbox:~/go/src/k8s.io/kubernetes$ grep -i forbidden /opt/stack/logs/*.log | grep -i forbidden | grep system:serviceaccount | cut -f 2- -d ']' | sort | uniq -c
    134  Error monitoring node status: nodes is forbidden: User "system:serviceaccount:kube-system:cloud-node-controller" cannot list nodes at the cluster scope
    652  k8s.io/kubernetes/pkg/controller/cloud/pvlcontroller.go:110: Failed to list *v1.PersistentVolume: persistentvolumes is forbidden: User "system:serviceaccount:kube-system:pvl-controller" cannot list persistentvolumes at the cluster scope
      1  nodes "bigbox" is forbidden: User "system:serviceaccount:kube-system:cloud-node-controller" cannot get nodes at the cluster scope
```

the command line looks like this:

```
/home/dims/go/src/k8s.io/kubernetes/_output/local/bin/linux/amd64/hyperkube controller-manager --v=10 --vmodule= --service-account-private-key-file=/tmp/kube-serviceaccount.key --root-ca-file=/var/run/kubernetes/server-ca.crt --cluster-signing-cert-file=/var/run/kubernetes/server-ca.crt --cluster-signing-key-file=/var/run/kubernetes/server-ca.key --enable-hostpath-provisioner=true --pvclaimbinder-sync-period=15s --feature-gates=AllAlpha=false --cloud-provider=external --external-cloud-volume-plugin=openstack --cloud-config=/etc/kubernetes/cloud-config --kubeconfig /var/run/kubernetes/controller.kubeconfig --use-service-account-credentials --controllers=* --master=https://localhost:6443
```

### PR DESCRIPTION
**Pull request by dims**
**What this PR does / why we need it**:

CCM has a couple of ClusterRoleBinding/ClusterRole(s) that it needs
for cloud-node-controller, pvl-controller and shared-informers service
accounts.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #59943

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by dims** [comment]
Also see earlier somewhat related discussion in https://github.com/kubernetes/kubernetes/pull/56709 about `aws-cloud-provider`

**Comment by dims** [comment]
`pvl-controller` is from here: 
https://github.com/kubernetes/kubernetes/blob/master/cmd/cloud-controller-manager/app/controllermanager.go#L218

references to `shared-informers` are in two locations:
https://github.com/kubernetes/kubernetes/blob/master/cmd/cloud-controller-manager/app/controllermanager.go#L204
https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L405

references to `cloud-node-controller` are in two locations:
https://github.com/kubernetes/kubernetes/blob/master/cmd/cloud-controller-manager/app/controllermanager.go#L210
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cloud/node_controller.go#L90

**Comment by liggitt** [comment]
rootClientBuilder (L204) is not creating a service account named "shared-informers", it just adds a user agent to the requests, and makes them using the credentials the cloud controller manager was started with. I'd expect the same to be done at L405.

The general disposition of things inside cloud-controller-manager is unclear to me. I thought the point of splitting it into a separate process was that it would be extended by things out of tree and run separately. As such, I wasn't expecting the apiserver to bootstrap RBAC permissions for it.

**Comment by dims** [comment]
ack @liggitt i'll remove the reference to `shared-informers` and check again

**Comment by dims** [comment]
@liggitt guess i feel CCM is special, since we are telling people that they can use that while folks are transitioning to separate repos / binaries / daemonsets.

/sig cluster-lifecycle
/area cloudprovider

**Comment by liggitt** [comment]
> @liggitt guess i feel CCM is special, since we are telling people that they can use that while folks are transitioning to separate repos / binaries / daemonsets.

don't we want to start modeling what it looks like to run separately? I thought that would look like a manifest that would deploy the CCM component and set up the required permissions.

**Comment by dims** [comment]
/hold

i'll submit another variation

**Comment by andrewsykim** [comment]
We had a similar discussion in https://github.com/kubernetes/kubernetes/pull/53511

**Review by jhorwit2 — CHANGES_REQUESTED** [review]
I think the RBAC files go [here](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/rbac) instead.

**Comment by jhorwit2** [comment]
> The general disposition of things inside cloud-controller-manager is unclear to me. I thought the point of splitting it into a separate process was that it would be extended by things out of tree and run separately. As such, I wasn't expecting the apiserver to bootstrap RBAC permissions for it (since the extensions would likely include different permissions, depending on the cloud)

@liggitt how do you want to handle the service controller which already exists in core since it's part of the KCM currently? Should we bundle it with this even though it's part of the api server bootstrap?

**Comment by dims** [comment]
@jhorwit2 Done!

**Comment by dims** [comment]
fyi, https://kubernetes.io/docs/concepts/architecture/cloud-controller/#plugin-mechanism seems to have way lot more. i am running kubetest with local-up-cluster/ccm/openstack and added only those that showed up in the logs.

**Comment by dims** [comment]
answering some questions from @cheftako 
```
Hi, sorry your PR is caught in the discussion on the best way to set up the SA/RBAC data. Couple of quick questions. What do we need the CCM in local cluster? What if the api server isn't up when start_cloud_controller_manager runs? I assume the cluster/addons will cause this to be created in any cluster running the static add on-manager. Is there a reasons the local cluster isn't running that add on manager? Do we intend to create the SA/RBAC in clusters which are not (yet) running the CCM?
```

1) need ccm in local cluster for me to run a CI job for testing openstack
2) local-up-cluster.sh has a wait for api server to start and bails out if there is a problem
3) i removed the explicit kubectl for the yaml files from local-up-cluster, will leave it to the addon
4) see #3, so we let something else (or addon) to create the SA/RBAC

**Comment by dims** [comment]
@jhorwit2 i added the stuff from https://kubernetes.io/docs/concepts/architecture/cloud-controller/#plugin-mechanism  to my yaml file

**Comment by liggitt** [comment]
this LGTM

File: cluster/addons/rbac/cloud-controller-manager-roles.yaml
Comment (jhorwit2) [inline cluster/addons/rbac/cloud-controller-manager-roles.yaml]: A lot of these permissions are for when we don't use service accounts per controller. The CCM itself needs a subset of these. If i'm not mistaken it just needs get/list/watch for resources (informers), SA creation, events and endpoints.

**Review by jhorwit2 — CHANGES_REQUESTED** [review]

File: cluster/addons/rbac/cloud-controller-manager-roles.yaml
Comment (liggitt) [reply cluster/addons/rbac/cloud-controller-manager-roles.yaml]: in my opinion, the role should include the permissions you'd need to run the control loop in a pod

Comment (jhorwit2) [reply cluster/addons/rbac/cloud-controller-manager-roles.yaml]: @liggitt So you're saying the root service account should also have all the permissions of all the controllers?

File: cluster/addons/rbac/cloud-controller-manager-role-bindings.yaml
Comment (liggitt) [inline cluster/addons/rbac/cloud-controller-manager-role-bindings.yaml]: what runs the cloud controller manager as this service account?

File: cluster/addons/rbac/cloud-controller-manager-roles.yaml
Comment (liggitt) [reply cluster/addons/rbac/cloud-controller-manager-roles.yaml]: I agree with @jhorwit2... a controller manager should have permissions needed to launch its control loops. that typically means:
* service account setup permissions (read/write service accounts, read access to secrets)
* shared informer permissions (read access to relevant resources)

and that's about it

Comment (cheftako) [reply cluster/addons/rbac/cloud-controller-manager-roles.yaml]: Interestingly I found an orthogonal set of permissions needed to successfully start the CCM. I think rather than having a static list/yaml files in cluster/addons, we would be better of having a list/yaml files in a CP directory and the CP start script should package/copy the list/yaml files into cluster/addons. This also seems like it will more naturally work toward our goal state of allowing each CP to control things like what controllers they run and how many CMs are used to run those controller. It also means the CPs can sign in to having extra permissions as they get things like CM online. So for instance as most CPs don't yet have a CCM running they won't need to add these permissions yet. When they do add them they can better control the permissions given out.

**Comment by thockin** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/59945)

**Comment by dims** [comment]
/close

moved it to the openstack cloud provider repo

**Pull request closed without merging**
